core/src/main/java/com/datastax/oss/driver/internal/core/auth/RuntimePlainTextAuthProvider.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.auth;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import net.jcip.annotations.ThreadSafe;
+
+/**
+ * A simple authentication provider that supports SASL authentication using the PLAIN mechanism for
+ * version 3 (or above) of the CQL native protocol.
+ *
+ * <p>To activate this provider, you will need to define it at runtime, and pass it into the
+ * SessionBuilder during session creation.
+ *
+ * <pre>
+ *     SessionBuilder builder =
+ *         SessionUtils.baseBuilder()
+ *             .withAuth(new RuntimePlainTextAuthProvider("logPrefix", "username", "password"));
+ * </pre>
+ */
+@ThreadSafe
+public class RuntimePlainTextAuthProvider extends PlainTextAuthProviderBase {
+  private final String username;
```

Comment:
Let's apply JAVA-2306 here and avoid storing sensitive data as strings. I suggest that you create a `Credentials` instance upfront and store it here instead of these two fields.

Reply:
:+1: that's a good idea.

Reply:
So while this seemed like a good idea, it won't actually work. The credentials are cleared when used, and the auth provider get credentials is invoked a few times during session creation. As a result the credentials end up cleared as part of protocol negotiation. In order to accomplish this we would need a way to encrypt the user name password in memory, and recreate the credentials when needed.

Reply:
Then you could implement a `Credentials.copy()` method as follows:

```
public Credentials copy() {
  return new Credentials(username.clone(), password.clone());
}
```
And use it as below:
```
public class RuntimePlainTextAuthProvider extends PlainTextAuthProviderBase {
  private final Credentials credentials;

  public RuntimePlainTextAuthProvider(String username, String password) {
    this("", username, password);
  }

  public RuntimePlainTextAuthProvider(String logPrefix, String username, String password) {
    super(logPrefix);
    credentials = new Credentials(username.toCharArray(), password.toCharArray());
  }

  @NonNull
  @Override
  protected Credentials getCredentials() {
    return credentials.copy();
  }
}
```

Reply:
But then we're still holding the credentials in memory, because the original field is never getting cleared.
I don't think there's a way to avoid it when the credentials have been provided programmatically (unless we encrypted them, but that's way off topic). I'll add a few explanations in the javadocs.